### PR TITLE
fix CUDA device link step

### PR DESCRIPTION
**Pull request by fwyzard**

### CODE REVIEW

**Comment by fwyzard** [comment]
@cmsbuild please test

File: SCRAM/GMake/Makefile.rules [hunk 1]
```diff
@@ -646,8 +646,8 @@ endef
 define link_cuda_objs
   @$(startlog_$(1)) [ -d $(@D) ] ||  $(CMD_mkdir) -p $(@D) &&\
   $(CMD_echo) ">> Cuda Device Link $@ " &&\
-  $(VERB_ECHO) $(SCRAM_PREFIX_COMPILER_COMMAND) nvcc -dlink $(call AdjustFlags,$1,,CPPFLAGS) --compiler-options "$(CUDA_CFLAGS)" $(CUDA_FLAGS) $? -o $@ &&\
-              ($(SCRAM_PREFIX_COMPILER_COMMAND) nvcc -dlink $(call AdjustFlags,$1,,CPPFLAGS) --compiler-options "$(CUDA_CFLAGS)" $(CUDA_FLAGS) $? -o $@) $(endlog_$(1))
+  $(VERB_ECHO) $(SCRAM_PREFIX_COMPILER_COMMAND) $(NVCC) -dlink $(call AdjustFlags,$1,,CPPFLAGS) --compiler-options "$(CUDA_CFLAGS)" $(CUDA_FLAGS) $? -o $@ &&\
+              ($(SCRAM_PREFIX_COMPILER_COMMAND) $(NVCC) -dlink $(call AdjustFlags,$1,,CPPFLAGS) --compiler-options "$(CUDA_CFLAGS)" $(CUDA_FLAGS) $? -o $@) $(endlog_$(1))
```
Comment (fwyzard) [inline SCRAM/GMake/Makefile.rules]: I don't know why, but this lets `nvcc` pick up the correct setup

File: SCRAM/GMake/Makefile.rules [hunk 2]
```diff
@@ -1136,14 +1136,15 @@ $(eval $(call LogFile,$(1),$(2)))
 .PHONY: $(1) all_$(1)
 $(1)_objdir := $(WORKINGDIR)/$(2)/$(1)
 $(1)_srcdir       := $(LOCALTOP)/$(2)
-ifneq ($(strip $(CUDASRC_FILES_SUFFIXES)),)
 $(1)_cudafiles    := $$(filter $(foreach e,$(CUDASRC_FILES_SUFFIXES),%.$e),$($(1)_files))
+ifneq ($(strip $$($(1)_cudafiles)),)
```
Comment (fwyzard) [inline SCRAM/GMake/Makefile.rules]: `CUDASRC_FILES_SUFFIXES` is always defined, this way `$(1)_cudaobj` is defined only if there is any .cu files

File: SCRAM/GMake/Makefile.rules [hunk 3]
```diff
@@ -1136,14 +1136,15 @@ $(eval $(call LogFile,$(1),$(2)))
...
 endif
-$(1)_files        := $$(filter-out $($(1)_SKIP_FILES) $$($(1)_cudafiles), $($(1)_files))
-$(1)_prodsrc      := $$(addprefix $(patsubst src/%,%,$2),$$($(1)_files) $($(1)_cudafiles))
-$(1)_files_exts   := $$(sort $$(patsubst .%,%,$$(suffix $$($(1)_files) $$($(1)_cudafiles))))
-$(1)_LOC_USE      += $$(if $$(strip $$(filter $(FORTRANSRC_FILES_SUFFIXES),$$($(1)_files_exts))),$(F77_TYPE_COMPILER),) $(if $(strip $(CUDASRC_FILES_SUFFIXES)),$$(if $$(strip $$(filter $(CUDASRC_FILES_SUFFIXES),$$($(1)_files_exts))),$(CUDA_TYPE_COMPILER),))
+$(1)_files        := $$(filter-out $($(1)_SKIP_FILES), $($(1)_files))
+$(1)_prodsrc      := $$(addprefix $(patsubst src/%,%,$2),$$($(1)_files))
+$(1)_files_exts   := $$(sort $$(patsubst .%,%,$$(suffix $$($(1)_files))))
+$(1)_LOC_USE      += $$(if $$(strip $$(filter $(FORTRANSRC_FILES_SUFFIXES),$$($(1)_files_exts))),$(F77_TYPE_COMPILER),)
+$(1)_LOC_USE      += $$(if $$(strip $$(filter $(CUDASRC_FILES_SUFFIXES),$$($(1)_files_exts))),$(CUDA_TYPE_COMPILER),)
```
Comment (fwyzard) [inline SCRAM/GMake/Makefile.rules]: I split this in two and made the CUDA case similar to the fortran case - but I do not know what this does...

File: SCRAM/GMake/Makefile.rules [hunk 4]
```diff
@@ -1136,14 +1136,15 @@ $(eval $(call LogFile,$(1),$(2)))
...
-ifneq ($(strip $(CUDASRC_FILES_SUFFIXES)),)
 $(1)_cudafiles    := $$(filter $(foreach e,$(CUDASRC_FILES_SUFFIXES),%.$e),$($(1)_files))
+ifneq ($(strip $$($(1)_cudafiles)),)
 $(1)_cudaobj      := $$($(1)_objdir)/$(1)_scramcuda.$(OBJEXT)
 endif
-$(1)_files        := $$(filter-out $($(1)_SKIP_FILES) $$($(1)_cudafiles), $($(1)_files))
-$(1)_prodsrc      := $$(addprefix $(patsubst src/%,%,$2),$$($(1)_files) $($(1)_cudafiles))
-$(1)_files_exts   := $$(sort $$(patsubst .%,%,$$(suffix $$($(1)_files) $$($(1)_cudafiles))))
-$(1)_LOC_USE      += $$(if $$(strip $$(filter $(FORTRANSRC_FILES_SUFFIXES),$$($(1)_files_exts))),$(F77_TYPE_COMPILER),) $(if $(strip $(CUDASRC_FILES_SUFFIXES)),$$(if $$(strip $$(filter $(CUDASRC_FILES_SUFFIXES),$$($(1)_files_exts))),$(CUDA_TYPE_COMPILER),))
+$(1)_files        := $$(filter-out $($(1)_SKIP_FILES), $($(1)_files))
```
Comment (fwyzard) [inline SCRAM/GMake/Makefile.rules]: Keep the .cu files, the corresponding .o files need to be linked twice: once by nvcc (for the CUDA specific code) and once by g++ (for the system code)

File: SCRAM/GMake/Makefile.rules (hunk 1 above)
Comment (smuzaffar) [reply SCRAM/GMake/Makefile.rules]: thanks, I should have used $(NVCC) in the first place.

File: SCRAM/GMake/Makefile.rules (hunk 2 above)
Comment (smuzaffar) [reply SCRAM/GMake/Makefile.rules]: you are right, CUDASRC_FILES_SUFFIXES is always set

File: SCRAM/GMake/Makefile.rules (hunk 3 above)
Comment (smuzaffar) [reply SCRAM/GMake/Makefile.rules]: this adds automatic dependency on
- fortran (if we have .f, .f77 source files)
- cuda if we have .cu files

**Pull request merged**